# Progetto di  Applicazioni Data Intensive

Progetto svolto nell'anno accademico 2024-2025 da:
- Filippo Massari 00010714020
- Francesco Pazzaglia 0001077423

## Introduzione
I dati presi in esame riguardano lo stile di vita di circa 80.000 studenti includendo dati come: l'ambiente di studio, la salute mentale, background familiare, motivazione e fattori ambientali.

Il dataset è stato preso su [Kaggle](https://www.kaggle.com/datasets/aryan208/student-habits-and-academic-performance-dataset).

Ci poniamo come obiettivo quello di sfruttare i dati precedentemente descritti per ricavarne una predizione delle performance studentesche del soggetto preso in analisi. 

## Implementazione

### Caricamento delle librerie

Per prima cosa importiamo le librerie che ci serviranno per poter manipolare ed effettuare operazioni sui dati ottenuti all'interno del dataset. Tra cui:

- *NumPy* per l'elaborazione numerica in Python di array e matrici;

- *pandas* per l'importazione ed esportazione dei dati;

- *matplotlib* per la rappresentazione grafica dei risultati ottenuti;
    - importiamo il modulo di matplotlib inline per far in modo che gli output vengano renderizzati direttamente nel file Jupyter piuttosto che all'esterno
- *IPython.display* funzione utile per printare elementi in Jupyter Notebook poiché gestisce la visualizzazione di DataFrames in un formato tabellare preciso;

In [47]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from IPython.display import display

%matplotlib inline

### Caricamento dei dati

Carichiamo il dataset delle abitudini degli studenti.

In [48]:
import os.path
file = "./data/enhanced_student_habits_performance_dataset.csv"
if not os.path.exists(file):
    print("Missing datasets, retrievit from kaggle: https://www.kaggle.com/datasets/aryan208/student-habits-and-academic-performance-dataset")
    
students = pd.read_csv(file, index_col=0)

### Significato delle colonne del Dataset

TODO: SISTEMARE I CAMPI (NON TUTTI SONO PRESENTI)

| Variabile                         | Descrizione                                                                                             |
|:----------------------------------|:--------------------------------------------------------------------------------------------------------|
| **student_id**                    | Identificativo univoco dello studente                                                                   |
| **age**                           | Età dello studente (16–28 anni)                                                                         |
| **gender**                        | Genere: Maschile, Femminile o Altro                                                                     |
| **major**                         | Corso di laurea o area di studio (es. Informatica, Ingegneria, Arti)                                    |
| **study_hours_per_day**           | Ore medie di studio giornaliere                                                                        |
| **social_media_hours**            | Ore trascorse sui social media                                                                          |
| **netflix_hours**                 | Ore trascorse a guardare Netflix                                                                        |
| **screen_time**                   | Tempo complessivo trascorso davanti agli schermi                                                         |
| **part_time_job**                 | Se lo studente ha un lavoro part-time (Sì/No)                                                           |
| **attendance_percentage**         | Percentuale di frequenza alle lezioni                                                                   |
| **sleep_hours**                   | Ore di sonno giornaliere                                                                               |
| **exercise_frequency**            | Frequenza dell’attività fisica                                                                           |
| **diet_quality**                  | Qualità della dieta                                                                                     |
| **mental_health_rating**          | Valutazione dello stato di salute mentale (1–10)                                                        |
| **stress_level**                  | Livello di stress (1–10)                                                                                 |
| **exam_anxiety_score**            | Punteggio di ansia da esame (1–10)                                                                      |
| **extracurricular_participation** | Partecipazione ad attività extracurriculari                                                             |
| **access_to_tutoring**            | Accesso a servizi di tutoraggio                                                                         |
| **family_income_range**           | Fascia di reddito familiare                                                                             |
| **parental_support_level**        | Livello di supporto da parte dei genitori                                                               |
| **parental_education_level**      | Livello di istruzione dei genitori                                                                      |
| **motivation_level**              | Livello di motivazione (1–10)                                                                           |
| **time_management_score**         | Punteggio di capacità di gestione del tempo (1–10)                                                      |
| **learning_style**                | Stile di apprendimento preferito                                                                         |
| **study_environment**             | Ambiente abituale di studio                                                                              |
| **dropout_risk**                  | Rischio di abbandono degli studi (Sì/No) — derivato da stress e motivazione                              |
| **previous_gpa**                  | Media dei voti precedenti                                                                                |
| **exam_score**                    | **Valore da predire: punteggio all’esame**                                                               |


In [49]:
students

age  \
student_id        
100000       26   
100001       28   
100002       17   
100003       27   
100004       25   
...         ...   
179995       16   
179996       16   
179997       26   
179998       23   
179999       25   

            gender  \
student_id           
100000        Male   
100001        Male   
100002        Male   
100003       Other   
100004      Female   
...            ...   
179995        Male   
179996      Female   
179997      Female   
179998       Other   
179999       Other   

                       major  \
student_id                     
100000      Computer Science   
100001                  Arts   
100002                  Arts   
100003            Psychology   
100004              Business   
...                      ...   
179995           Engineering   
179996              Business   
179997                  Arts   
179998               Biology   
179999           Engineering   

            study_hours_per_day  \
student_id                        
100000                 7.645367   
100001                 5.700000   
100002                 2.400000   
100003                 3.400000   
100004                 4.700000   
...                         ...   
179995                 3.700000   
179996                 1.200000   
179997                 4.100000   
179998                 3.800000   
179999                 3.000000   

            social_media_hours  \
student_id                       
100000                     3.0   
100001                     0.5   
100002                     4.2   
100003                     4.6   
100004                     0.8   
...                        ...   
179995                     2.1   
179996                     0.4   
179997                     1.6   
179998                     0.6   
179999                     4.5   

            netflix_hours  \
student_id                  
100000                0.1   
100001                0.4   
100002                0.7   
100003                2.3   
100004                2.7   
...                   ...   
179995                1.0   
179996                2.9   
179997                1.6   
179998                3.5   
179999                3.4   

           part_time_job  \
student_id                 
100000               Yes   
100001                No   
100002                No   
100003               Yes   
100004               Yes   
...                  ...   
179995               Yes   
179996                No   
179997               Yes   
179998               Yes   
179999               Yes   

            attendance_percentage  \
student_id                          
100000                       70.3   
100001                       88.4   
100002                       82.1   
100003                       79.3   
100004                       62.9   
...                           ...   
179995                       80.8   
179996                       99.5   
179997                       46.1   
179998                       58.7   
179999                       97.7   

            sleep_hours  \
student_id                
100000              6.2   
100001              7.2   
100002              9.2   
100003              4.2   
100004              6.5   
...                 ...   
179995              6.1   
179996              4.1   
179997              8.3   
179998              5.8   
179999              5.3   

           diet_quality  \
student_id                
100000             Poor   
100001             Good   
100002             Good   
100003             Fair   
100004             Good   
...                 ...   
179995             Fair   
179996             Good   
179997             Good   
179998             Good   
179999             Fair   

            ...  \
student_id  ...   
100000      ...   
100001      ...   
100002      ...   
100003      ...   
100004      ...   
...         ...   
179995      ...   
179996      ...   
179997      ...   
179998      ...   
1799

In [50]:
students.info(verbose=False, memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
Index: 80000 entries, 100000 to 179999
Columns: 30 entries, age to exam_score
dtypes: float64(10), int64(8), object(12)
memory usage: 68.4 MB


Definiamo l’elenco delle colonne da trattare come categorical e al caricamento ne specifichiamo il tipo. Andiamo infine a ricalcolare la dimensione del file costatando una netta diminuzione dello spazio occupato rispetto all'originale.

In [51]:
cols_to_cat = [
    'gender', 'major', 'part_time_job', 'diet_quality',
    'study_environment', 'access_to_tutoring',
    'family_income_range', 'parental_support_level',
    'parental_education_level', 'learning_style'
]

dtype = {col: 'category' for col in cols_to_cat}
students = pd.read_csv(file, dtype=dtype)

for col in cols_to_cat:
    students[col] = students[col].astype('category')

students.info(verbose=False, memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80000 entries, 0 to 79999
Columns: 31 entries, student_id to exam_score
dtypes: category(10), float64(10), int64(8), object(3)
memory usage: 25.5 MB


### Preparazione dati

Riteniamo che alcune colonne del dataset siano similabili per similarità dei dati forniti, quindi procediamo ad accorparle per attinenza semantica. 

TODO: screen_productivity_hours SEMBRA ESSERE ERRATO

In [52]:
students["screen_entertainment_hours"] = students["social_media_hours"] + students["netflix_hours"]
students.drop(columns=["social_media_hours", "netflix_hours"], inplace=True)
ent = students.pop("screen_entertainment_hours")
students.insert(loc=5, column="screen_entertainment_hours", value=ent)

 TODO DESCRIZIONE

In [53]:
students["screen_productivity_hours"] = students["screen_time"] - students["screen_entertainment_hours"]
prod = students.pop("screen_productivity_hours")
students.insert(loc=6, column="screen_productivity_hours", value=prod)

In [54]:
students.head(1)

student_id  \
0      100000   

   age  \
0   26   

  gender  \
0   Male   

              major  \
0  Computer Science   

   study_hours_per_day  \
0             7.645367   

   screen_entertainment_hours  \
0                         3.1   

   screen_productivity_hours  \
0                        7.8   

  part_time_job  \
0           Yes   

   attendance_percentage  \
0                   70.3   

   sleep_hours  \
0          6.2   

   ...  \
0  ...   

  screen_time  \
0        10.9   

   study_environment  \
0  Co-Learning Group   

  access_to_tutoring  \
0                Yes   

  family_income_range  \
0                High   

   parental_support_level  \
0                       9   

  motivation_level  \
0                7   

   exam_anxiety_score  \
0                   8   

   learning_style  \
0         Reading   

   time_management_score  \
0                    3.0   

  exam_score  
0        100  

[1 rows x 31 columns]

In [55]:
category_order = [
    "Identifiers",
    "Demographics",
    "Academic Engagement",
    "Lifestyle & Health",
    "Screen Activity",
    "Family & Socioeconomic Support",
    "Personal Skills",
    "Behavior & Preferences",
    "Outcomes"
]


category_map = {
    "student_id": "Identifiers",

    "age":       "Demographics",
    "gender":    "Demographics",
    "major":     "Demographics",

    "semester":                      "Academic Engagement",
    "study_hours_per_day":           "Academic Engagement",
    "attendance_percentage":         "Academic Engagement",
    "access_to_tutoring":            "Academic Engagement",
    "extracurricular_participation": "Academic Engagement",
    "previous_gpa":                  "Academic Engagement",


    "sleep_hours":          "Lifestyle & Health",
    "diet_quality":         "Lifestyle & Health",
    "exercise_frequency":   "Lifestyle & Health",
    "mental_health_rating": "Lifestyle & Health",
    "stress_level":         "Lifestyle & Health",
    "exam_anxiety_score":   "Lifestyle & Health",

    "screen_time":                "Screen Activity",
    "screen_productivity_hours":  "Screen Activity",
    "screen_entertainment_hours": "Screen Activity",
    "social_activity":            "Screen Activity",

    "family_income_range":      "Family & Socioeconomic Support",
    "parental_education_level": "Family & Socioeconomic Support",
    "parental_support_level":   "Family & Socioeconomic Support",
    "internet_quality":         "Family & Socioeconomic Support",
    
    
    "time_management_score": "Personal Skills",
    "motivation_level":      "Personal Skills",
    "learning_style":        "Personal Skills",

    
    "part_time_job":      "Behavior & Preferences",
    "dropout_risk":       "Behavior & Preferences",
    "study_environment":  "Behavior & Preferences",

    "exam_score":   "Outcomes"
}


In [56]:
missing_cols = [col for col in students.columns if col not in category_map]
if missing_cols:
    raise ValueError(f"The following columns are not present in category_map: {missing_cols}")

cols_sorted = sorted(
    students.columns,
    key=lambda col: category_order.index(category_map[col])
)

students = students[cols_sorted]

In [57]:
print(students.columns.tolist())

['student_id', 'age', 'gender', 'major', 'study_hours_per_day', 'attendance_percentage', 'extracurricular_participation', 'previous_gpa', 'semester', 'access_to_tutoring', 'sleep_hours', 'diet_quality', 'exercise_frequency', 'mental_health_rating', 'stress_level', 'exam_anxiety_score', 'screen_entertainment_hours', 'screen_productivity_hours', 'social_activity', 'screen_time', 'parental_education_level', 'internet_quality', 'family_income_range', 'parental_support_level', 'motivation_level', 'learning_style', 'time_management_score', 'part_time_job', 'dropout_risk', 'study_environment', 'exam_score']


### ANALISI DEI DATI


Procediamo poi a ricavare alcune statistiche descrittive sui dati degli studenti. Questo approccio raffina l'analisi visualizzando le statistiche necessarie per interpretare il rendimento accademico e le abitudini degli studenti. In particolar modo prendaimo in analisi la media, il massimo ed il miniomo dei valori presenti nelle principali colonne.


In [58]:
studentDescribed = students.describe()

rows_to_drop = ['count', 'std', '25%', '50%', '75%']
filtered_studentDescribed = studentDescribed.drop(index=rows_to_drop)

filtered_studentDescribed = filtered_studentDescribed.drop(columns=["student_id"])

segment_size = 5 

for start in range(0, len(filtered_studentDescribed.columns), segment_size):
    stop = start + segment_size
    display(filtered_studentDescribed.iloc[:, start:stop])

age  \
mean  22.004288   
min   16.000000   
max   28.000000   

      study_hours_per_day  \
mean             4.174388   
min              0.000000   
max             12.000000   

      attendance_percentage  \
mean              69.967884   
min               40.000000   
max              100.000000   

      previous_gpa  \
mean      3.602448   
min       1.640000   
max       4.000000   

      semester  
mean  4.497338  
min   1.000000  
max   8.000000

sleep_hours  \
mean     7.017417   
min      4.000000   
max     12.000000   

      exercise_frequency  \
mean            3.516587   
min             0.000000   
max             7.000000   

      mental_health_rating  \
mean              6.804107   
min               1.000000   
max              10.000000   

      stress_level  \
mean      5.012478   
min       1.000000   
max      10.000000   

      exam_anxiety_score  
mean            8.508475  
min             5.000000  
max            10.000000

screen_entertainment_hours  \
mean                     4.49912   
min                      0.00000   
max                      9.00000   

      screen_productivity_hours  \
mean               5.173909e+00   
min               -8.881784e-16   
max                1.380000e+01   

      social_activity  \
mean         2.500225   
min          0.000000   
max          5.000000   

      screen_time  \
mean     9.673029   
min      0.300000   
max     21.000000   

      motivation_level  
mean          5.488525  
min           1.000000  
max          10.000000

time_management_score  \
mean               5.499132   
min                1.000000   
max               10.000000   

      exam_score  
mean    89.14135  
min     36.00000  
max    100.00000

- **Age**: Gli studenti hanno un'età media di circa 22 anni, con un range tra i 16 e i 28 anni, tipico di un'ampia fascia universitaria.
- **Study Hours per Day**: Il numero medio di ore di studio è 4.17, suggerendo che la maggior parte degli studenti dedica un tempo moderato allo studio giornaliero. Tuttavia, un massimo di 12 ore indicano alcuni outlier che potrebbero mettere molta enfasi sulla preparazione accademica.
- **Attendance Percentage**: La frequenza media è del 69.97%, con un minimo del 40% e massimo del 100%. Ciò indica una discreta variabilità nella partecipazione alle lezioni, cruciale per la performance accademica.
- **Previous GPA**: Con una media di 3.60, gli studenti mostrano un buon rendimento accademico precedente, evidenziato anche dal "massimo" di 4.0, il che rappresenta di solito il voto più elevato raggiungibile.
- **Semester**: La media è di 4.5, suggerendo che gli studenti si trovano per lo più a metà del loro percorso accademico, tra il primo e l'ultimo semestre.
- **Sleep Hours**: Gli studenti dormono mediamente 7 ore, che è vicino al raccomandato per la loro età. Tuttavia, alcuni studenti dormono solo 4 ore, potenzialmente influenzando negativamente la loro salute mentale e rendimento.
- **Exercise Frequency**: Con una media di 3.5, i dati indicano che gli studenti fanno esercizio regolarmente un paio di volte a settimana; il minimo di 0 segnala una totale assenza di attività fisica tra alcuni studenti.
- **Mental Health Rating e Stress Level**: La salute mentale media è valutata a 6.8 su 10, mentre il livello di stress è mediamente 5 su 10. Questo equilibrio suggerisce che, sebbene gli studenti affrontino stress, mantengono una salute mentale relativamente positiva.
- **Exam Anxiety Score**: Il livello di ansia è elevato, con una media di 8.5, indicando che gli esami sono una fonte significativa di pressione, potenzialmente impattando sulle performance.
- **Screen Entertainment e Productivity Hours**: Gli studenti passano più di 4 ore al giorno in attività di intrattenimento e poco più di 5 in attività produttive; la gestione del tempo tra intrattenimento e produttività è cruciale per la riuscita accademica.
- **Social Activity**: Con una media bassa di 2.5, l'interazione sociale può essere limitante ed è opportuno valutare se impatti sul benessere generale.
- **Motivation Level**: La motivazione media è 5.4, suggerendo che metà degli studenti potrebbe mancare di guida o interesse nei loro studi, un fattore che potrebbe ridurre l'engagement accademico complessivo.
- **Time Management Score**: In media 5.5, suggerisce che la gestione del tempo è moderata, con possibilità di miglioramento per molti studenti.